src/DynamoCore/Models/DynamoModel.cs

```diff
@@ -412,6 +412,111 @@ public static DynamoModel Start(IStartConfiguration configuration)
+                }
+                else
+                {
+                    fileMinorPart--;
+                }
+                var previousVersionPath = Path.Combine(rootFolder, String.Format("{0}.{1}", fileMajorPart, fileMinorPart));
+                if (Directory.Exists(previousVersionPath))
+                    break;
+                
+            }
+            return new FileVersion(fileMajorPart, fileMinorPart);
+        }
+
+        private static DynamoMigratorBase CreateMigrator(string rootFolder, int majorVersion, int minorVersion)
+        {
+            var className = "Dynamo.Core.DynamoMigrator" + majorVersion + minorVersion;
+
+            var args = new object[] { rootFolder };
+            var type = Assembly.GetExecutingAssembly().GetType(className);
+            return (DynamoMigratorBase)Activator.CreateInstance(type, args);
+        }
+
+        // Reference: https://msdn.microsoft.com/en-us/library/system.io.directoryinfo.aspx
+        private static void Copy(string sourceDirectory, string targetDirectory)
+        {
+            var diSource = new DirectoryInfo(sourceDirectory);
+            var diTarget = new DirectoryInfo(targetDirectory);
+
+            CopyAll(diSource, diTarget);
+        }
+
+        private static void CopyAll(DirectoryInfo source, DirectoryInfo target)
+        {
+            // Check if the target directory exists; if not, create it.
+            if (Directory.Exists(target.FullName) == false)
+            {
+                Directory.CreateDirectory(target.FullName);
+            }
+
+            // Copy each file into the new directory.
+            foreach (FileInfo fi in source.GetFiles())
+            {
+                Console.WriteLine(@"Copying {0}\{1}", target.FullName, fi.Name);
+                fi.CopyTo(Path.Combine(target.FullName, fi.Name), true);
+            }
+
+            // Copy each subdirectory using recursion.
+            foreach (DirectoryInfo diSourceSubDir in source.GetDirectories())
+            {
+                DirectoryInfo nextTargetSubDir =
+                    target.CreateSubdirectory(diSourceSubDir.Name);
+                CopyAll(diSourceSubDir, nextTargetSubDir);
+            }
+        }
+
+        public PreferenceSettings MigrateBetweenDynamoVersions()
+        {
+            var assemblyPath = Assembly.GetExecutingAssembly().Location;
+            var currentVersionInfo = FileVersionInfo.GetVersionInfo(assemblyPath);
+
+            // No migration required if the current version is <= version 0.7
+            if (currentVersionInfo.FileMajorPart == 0 &&
+                currentVersionInfo.FileMinorPart <= 7)
+                return null;
+
+            // Get root folder - this will be different for DynamoPro
+            var rootFolder = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "Dynamo");
+            var previousVersion = FindPreviousVersion(rootFolder, currentVersionInfo);
```

Comment:
New property exposed in https://github.com/DynamoDS/Dynamo/pull/4006 `IPathManager.UserDataDirectory`, does that help?